core/drivers/stm32_rng.c

```diff
@@ -219,67 +372,171 @@ TEE_Result hw_get_random_bytes(void *out, size_t size)
+
+	assert(stm32_rng);
+
+	res = clk_enable(stm32_rng->clock);
+	if (res)
+		return res;
+
+	if (op == PM_OP_RESUME)
+		res = stm32_rng_pm_resume();
+	else
+		res = stm32_rng_pm_suspend();
+
+	clk_disable(stm32_rng->clock);
+
+	return res;
+}
+DECLARE_KEEP_PAGER(stm32_rng_pm);
+
 #ifdef CFG_EMBED_DTB
-static TEE_Result stm32_rng_init(void)
+static TEE_Result stm32_rng_parse_fdt(const void *fdt, int node)
 {
-	void *fdt = NULL;
-	int node = -1;
-	struct dt_node_info dt_info;
 	TEE_Result res = TEE_ERROR_GENERIC;
+	struct dt_node_info dt_rng = { };
 
-	memset(&dt_info, 0, sizeof(dt_info));
+	_fdt_fill_device_info(fdt, &dt_rng, node);
+	if (dt_rng.reg == DT_INFO_INVALID_REG)
+		return TEE_ERROR_BAD_PARAMETERS;
 
-	fdt = get_embedded_dt();
-	if (!fdt)
-		panic();
+	stm32_rng->base.pa = dt_rng.reg;
+	stm32_rng->base.va = io_pa_or_va_secure(&stm32_rng->base,
+						dt_rng.reg_size);
+	assert(stm32_rng->base.va);
 
-	while (true) {
-		node = fdt_node_offset_by_compatible(fdt, node, DT_RNG_COMPAT);
-		if (node < 0)
-			break;
+	res = rstctrl_dt_get_by_index(fdt, node, 0, &stm32_rng->rstctrl);
+	if (res != TEE_SUCCESS && res != TEE_ERROR_ITEM_NOT_FOUND)
+		return res;
 
-		_fdt_fill_device_info(fdt, &dt_info, node);
+	res = clk_dt_get_by_index(fdt, node, 0, &stm32_rng->clock);
+	if (res)
+		return res;
 
-		if (!(dt_info.status & DT_STATUS_OK_SEC))
-			continue;
+	if (fdt_getprop(fdt, node, "clock-error-detect", NULL))
+		stm32_rng->clock_error = true;
 
-		if (stm32_rng)
-			panic();
+	/* Release device if not used at runtime or for pm transitions */
+	stm32_rng->release_post_boot = IS_ENABLED(CFG_WITH_SOFTWARE_PRNG) &&
+				       !IS_ENABLED(CFG_PM);
 
-		stm32_rng = calloc(1, sizeof(*stm32_rng));
-		if (!stm32_rng)
-			panic();
+	return TEE_SUCCESS;
+}
 
-		assert(dt_info.clock != DT_INFO_INVALID_CLOCK &&
-		       dt_info.reg != DT_INFO_INVALID_REG &&
-		       dt_info.reg_size != DT_INFO_INVALID_REG_SIZE);
+static TEE_Result stm32_rng_probe(const void *fdt, int offs,
+				  const void *compat_data __unused)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
 
-		if (dt_info.status & DT_STATUS_OK_NSEC) {
-			stm32mp_register_non_secure_periph_iomem(dt_info.reg);
-			stm32_rng->release_post_boot = true;
-		} else {
-			stm32mp_register_secure_periph_iomem(dt_info.reg);
-		}
+	/* Expect a single RNG instance */
+	assert(!stm32_rng);
 
-		stm32_rng->base.pa = dt_info.reg;
-		if (!io_pa_or_va_secure(&stm32_rng->base, dt_info.reg_size))
-			panic();
+	stm32_rng = calloc(1, sizeof(*stm32_rng));
+	if (!stm32_rng)
+		panic();
 
-		res = clk_dt_get_by_index(fdt, node, 0, &stm32_rng->clock);
-		if (res)
-			return res;
+	res = stm32_rng_parse_fdt(fdt, offs);
+	if (res)
+		goto err;
 
-		assert(stm32_rng->clock);
+	stm32_rng->ddata = compat_data;
+	assert(stm32_rng->ddata);
 
-		DMSG("RNG init");
-	}
+	res = clk_enable(stm32_rng->clock);
+	if (res)
+		goto err;
+
+	if (stm32_rng->rstctrl &&
+	    rstctrl_assert_to(stm32_rng->rstctrl, RNG_RESET_TIMEOUT_US))
+		goto err;
```

Comment:
set `res = TEE_ERROR_GENERIC;` and disable the clock.
ditto at line 496.

Reply:
Good spot